lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,948 @@
+
+        if os.path.exists(b_collection_path):
+            shutil.rmtree(b_collection_path)
+        os.makedirs(b_collection_path)
+
+        with tarfile.open(self.b_path, mode='r') as collection_tar:
+            files_member_obj = collection_tar.getmember('FILES.json')
+            with _tarfile_extract(collection_tar, files_member_obj) as files_obj:
+                files = json.loads(to_text(files_obj.read(), errors='surrogate_or_strict'))
+
+            _extract_tar_file(collection_tar, 'MANIFEST.json', b_collection_path, b_temp_path)
+            _extract_tar_file(collection_tar, 'FILES.json', b_collection_path, b_temp_path)
+
+            for file_info in files['files']:
+                file_name = file_info['name']
+                if file_name == '.':
+                    continue
+
+                if file_info['ftype'] == 'file':
+                    _extract_tar_file(collection_tar, file_name, b_collection_path, b_temp_path,
+                                      expected_hash=file_info['chksum_sha256'])
+                else:
+                    os.makedirs(os.path.join(b_collection_path, to_bytes(file_name, errors='surrogate_or_strict')))
+
+    def set_latest_version(self):
+        self.versions = set([self.latest_version])
+        self._get_metadata()
+
+    def _get_metadata(self):
+        if self._metadata:
+            return
+
+        n_collection_url = _urljoin(self.source, 'api', 'v2', 'collections', self.namespace, self.name, 'versions',
+                                    self.latest_version)
+        details = json.load(open_url(n_collection_url, validate_certs=self._validate_certs))
+        self._galaxy_info = details
+        self._metadata = details['metadata']
+
+    def _meets_requirements(self, version, requirements, parent):
+        """
+        Supports version identifiers can be '==', '!=', '>', '>=', '<', '<=', '*'. Each requirement is delimited by ','
+        """
+        op_map = {
+            '!=': operator.ne,
+            '==': operator.eq,
+            '=': operator.eq,
+            '>=': operator.ge,
+            '>': operator.gt,
+            '<=': operator.le,
+            '<': operator.lt,
+        }
+
+        for req in list(requirements.split(',')):
+            op_pos = 2 if len(req) > 1 and req[1] == '=' else 1
+            op = op_map.get(req[:op_pos])
+
+            requirement = req[op_pos:]
+            if not op:
+                requirement = req
+                op = operator.eq
+
+                # In the case we are checking a new requirement on a base requirement (parent != None) we can't accept
+                # version as '*' (unknown version) unless the requirement is also '*'.
+                if parent and version == '*' and requirement != '*':
+                    break
+                elif requirement == '*' or version == '*':
+                    continue
+
+            if not op(LooseVersion(version), LooseVersion(requirement)):
```

Comment:
The code right now just relies on Galaxy to handle the semver strictness. It doesn't validate that the version specified matches semvar and just relies on LooseVersion. This is mainly because I cannot rely on the semver package to be installed and this was the path of least resistance.

Reply:
So talked about this a bit more internally and the prevailing thought was we don't want to add any deps and want to make this as simple as possible. In the end we will not validate the versions but just leave it for Galaxy to deal manage. A potential issue of using LooseVersion that ordered a prerelease string (2.0.0-beta.1) as newer than the actual release (2.0.0) has been fixed in the latest commit. Any pre-release versions are filtered out unless an explicit requirement for that pre-release was set.